### DISCUSSION TITLE
Mini batch training of collocation points?

### DISCUSSION THREAD

**Discussion by mattragoza** [lululxvi/deepxde#744] [Q&A]
My question is how to do minibatch training with deepxde using point set boundary conditions?

I noticed in issue #320 that it is possible to use `PDEResidualResampler` to sample new PDE training points on each epoch. However, this is not sufficient for my use case. I am solving an inverse problem where the boundary conditions are defined by a large multidimensional displacement image (hundreds of thousands of vector-valued data points). I provide these to deepxde as a `PointSetBC` which works just fine for smaller problems, but I run out of memory when scaling up to the full problem.

I need a way to train using minibatches that resample not just the PDE residual points, but the boundary condition collocation points as well. Is this possible currently or do I need to implement a new trainer?

**Comment by lululxvi** [comment]
For any PDE, the number inside the domain for PDE loss is always orders of magnitude larger than the number of points in the boundary, right? If the problem has a lot of BC points, then the PDE points should be even more. So the PDE point is always dominant. You can simply do a downsampling for the BC points, unless I misunderstand your problem.

On the other hand, you can also check [PointCloud](https://deepxde.readthedocs.io/en/latest/modules/deepxde.geometry.html#deepxde.geometry.pointcloud.PointCloud)

**Reply by mattragoza** [reply]
Maybe describing my problem in more detail will help. I am trying to solve the following inverse Helmholtz problem:

$\mu \nabla^2 u = -\rho \omega^2 u$

Where  $u: R^3 \to C^3 $  represents 3D steady-state displacements and  $\mu: R^3 \to C $ represents a 3D map of stiffness. I have a set of observed displacement values in the form of a very large multidimensional image, and I would like to solve the PDE in order to reconstruct $\mu$. The displacement values are measured on a regular grid that spans the entire domain of interest, so I would prefer to just use these data points rather than sampling additional points in the domain for which the displacement values are not known.

I would like to use all of the observed displacement values for training the model instead of downsampling the image. Ideally, I would be able to do this by training the model on batches of image coordinates/displacement value pairs. I am currently providing these data pairs to the model using a `PointSetBC`, but I don't know how to resample or subsample these BC points to implement batch training.

**Comment by lululxvi** [comment]
`PointSetBC` can be used together with `PDEResidualResampler` to sample mini-batch of PDE residual points, in the same way as other standard geometry. But for the points of data loss, DeepXDE currently doesn't support mini-batch. You have to modify the code. But based on my experience, a down-sampling of u could be enough.

**Reply by mattragoza** [reply]
Thanks for your response. I was able to solve this by creating a subclass of `PointSetBC` that handles the resampling of the data points internally. Each call to `bc.collocation_points()` gets and returns the next batch of training points, then subsequent calls to `bc.error()` computes the boundary condition loss wrt the current batch.

**Reply by lululxvi** [reply]
Great! If you would like, feel free to submit a PR.
